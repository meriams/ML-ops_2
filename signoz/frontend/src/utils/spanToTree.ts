/* eslint-disable no-restricted-syntax */
import { cloneDeep } from 'lodash-es';
import { ITraceForest, ITraceTree, Span } from 'types/api/trace/getTraceItem';

const getSpanReferences = (
	rawReferences: string[] = [],
	isChildReference: boolean,
): Record<string, string>[] => {
	let filteredReferences = [];
	if (isChildReference) {
		filteredReferences = rawReferences.filter((value) =>
			value.includes('CHILD_OF'),
		);
	} else {
		filteredReferences = rawReferences.filter(
			(value) => !value.includes('CHILD_OF'),
		);
	}

	return filteredReferences.map((rawRef) => {
		const refObject: Record<string, string> = {};
		rawRef
			.replaceAll('{', '')
			.replaceAll('}', '')
			.replaceAll(' ', '')
			.split(',')
			.forEach((rawRefKeyPair) => {
				const [key, value] = rawRefKeyPair.split('=');
				refObject[key] = value;
			});

		return refObject;
	});
};

// This getSpanTags is migrated from the previous implementation.
const getSpanTags = (spanData: Span): { key: string; value: string }[] => {
	const tags = [];
	if (spanData[7] !== null && spanData[8] !== null) {
		if (typeof spanData[7] === 'string' && typeof spanData[8] === 'string') {
			tags.push({ key: spanData[7], value: spanData[8] });
		} else if (spanData[7].length > 0 && spanData[8].length > 0) {
			for (let j = 0; j < spanData[7].length; j += 1) {
				tags.push({ key: spanData[7][j], value: spanData[8][j] });
			}
		}
	}
	return tags;
};

// eslint-disable-next-line sonarjs/cognitive-complexity
export const spanToTreeUtil = (inputSpanList: Span[]): ITraceForest => {
	const spanList = cloneDeep(inputSpanList);
	const traceIdSet: Set<string> = new Set();
	const spanMap: Record<string, ITraceTree> = {};

	const createTraceRootSpan = (
		spanReferences: Record<string, string>[],
	): void => {
		spanReferences.forEach(({ SpanId, TraceId }) => {
			traceIdSet.add(TraceId);
			if (SpanId && !spanMap[SpanId]) {
				spanMap[SpanId] = {
					id: SpanId,
					name: `Missing Span (${SpanId})`,
					children: [],
					serviceColour: '',
					serviceName: '',
					startTime: null as never,
					tags: [],
					time: null as never,
					value: null as never,
					isMissing: true,
				};
			}
		});
	};

	spanList.forEach((span) => {
		const childReferences = getSpanReferences(span[9] as string[], true);
		const nonChildReferences = getSpanReferences(span[9] as string[], false);
		const spanObject = {
			id: span[1],
			name: span[4],
			value: parseInt(span[6], 10),
			time: parseInt(span[6], 10),
			startTime: span[0],
			tags: getSpanTags(span),
			children: [],
			serviceName: span[3],
			hasError: !!span[11],
			serviceColour: '',
			event: span[10]?.map((e) => JSON.parse(e || '{}') || {}),
			childReferences,
			nonChildReferences,
		};
		spanMap[span[1]] = spanObject;
	});

	for (const [, spanData] of Object.entries(spanMap)) {
		if (spanData.childReferences) {
			createTraceRootSpan(spanData.childReferences);
			spanData.childReferences.forEach(({ SpanId: parentSpanId }) => {
				if (spanMap[parentSpanId]) {
					spanData.isProcessed = true;
					spanMap[parentSpanId].children.push(spanData);
				}
			});
		}
	}
	for (const [spanId, spanData] of Object.entries(spanMap)) {
		if (spanData.isProcessed) {
			delete spanMap[spanId];
		}
	}

	const spanTree: ITraceTree[] = [];
	const missingSpanTree: ITraceTree[] = [];
	const referencedTraceIds: string[] = Array.from(traceIdSet);
	Object.keys(spanMap).forEach((spanId) => {
		const isRoot = spanMap[spanId].childReferences?.some(
			(refs) => refs.SpanId === '',
		);
		if (isRoot) {
			spanTree.push(spanMap[spanId]);
			return;
		}

		for (const traceId of referencedTraceIds) {
			if (traceId.includes(spanId)) {
				spanTree.push(spanMap[spanId]);
			} else {
				missingSpanTree.push(spanMap[spanId]);
			}
		}
	});

	return {
		spanTree,
		missingSpanTree,
	};
};
