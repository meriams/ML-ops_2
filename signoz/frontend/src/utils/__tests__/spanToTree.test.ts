import { TraceData } from '../fixtures/TraceData';
import { spanToTreeUtil } from '../spanToTree';

describe('utils/spanToTree', () => {
	test('should return a single tree on valid trace data', () => {
		const spanTree = spanToTreeUtil(TraceData);
		expect(spanTree.spanTree.length).toBe(1);
		expect(spanTree.missingSpanTree.length).toBe(0);
		expect(spanTree).toMatchSnapshot();
	});
	test('should return a single tree on valid trace data', () => {
		const MissingTraceData = [...TraceData];
		MissingTraceData.splice(1, 1);

		const spanTree = spanToTreeUtil(MissingTraceData);

		expect(spanTree.spanTree.length).toBe(1);
		expect(spanTree.missingSpanTree.length).toBe(1);
		expect(spanTree).toMatchSnapshot();
	});
});
