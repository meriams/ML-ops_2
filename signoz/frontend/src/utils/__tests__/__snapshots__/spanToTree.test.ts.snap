// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/spanToTree should return a single tree on valid trace data 1`] = `
Object {
  "missingSpanTree": Array [],
  "spanTree": Array [
    Object {
      "childReferences": Array [
        Object {
          "RefType": "CHILD_OF",
          "SpanId": "",
          "TraceId": "0000000000000000span_1",
        },
      ],
      "children": Array [
        Object {
          "childReferences": Array [
            Object {
              "RefType": "CHILD_OF",
              "SpanId": "span_1",
              "TraceId": "0000000000000000span_1",
            },
          ],
          "children": Array [
            Object {
              "childReferences": Array [
                Object {
                  "RefType": "CHILD_OF",
                  "SpanId": "span_2",
                  "TraceId": "0000000000000000span_1",
                },
              ],
              "children": Array [],
              "event": Array [
                Object {
                  "attributeMap": Object {
                    "event": "HTTP request received S3",
                    "level": "info",
                    "method": "GET",
                    "url": "/dispatch?customer=392&nonse=0.015296363321630757",
                  },
                  "timeUnixNano": 1657275433246142000,
                },
              ],
              "hasError": false,
              "id": "span_3",
              "isProcessed": true,
              "name": "HTTP GET SPAN 3",
              "nonChildReferences": Array [],
              "serviceColour": "",
              "serviceName": "frontend",
              "startTime": 1657275433246,
              "tags": Array [
                Object {
                  "key": "host.name.span3",
                  "value": "span_3",
                },
              ],
              "time": 683273000,
              "value": 683273000,
            },
          ],
          "event": Array [
            Object {
              "attributeMap": Object {
                "event": "HTTP request received S2",
                "level": "info",
                "method": "GET",
                "url": "/dispatch?customer=392&nonse=0.015296363321630757",
              },
              "timeUnixNano": 1657275433246142000,
            },
          ],
          "hasError": false,
          "id": "span_2",
          "isProcessed": true,
          "name": "HTTP GET SPAN 2",
          "nonChildReferences": Array [],
          "serviceColour": "",
          "serviceName": "frontend",
          "startTime": 1657275433246,
          "tags": Array [
            Object {
              "key": "host.name.span2",
              "value": "span_2",
            },
          ],
          "time": 683273000,
          "value": 683273000,
        },
      ],
      "event": Array [
        Object {
          "attributeMap": Object {
            "event": "HTTP request received S1",
            "level": "info",
            "method": "GET",
            "url": "/dispatch?customer=392&nonse=0.015296363321630757",
          },
          "timeUnixNano": 1657275433246142000,
        },
      ],
      "hasError": false,
      "id": "span_1",
      "name": "HTTP GET SPAN 1",
      "nonChildReferences": Array [],
      "serviceColour": "",
      "serviceName": "frontend",
      "startTime": 1657275433246,
      "tags": Array [
        Object {
          "key": "host.name.span1",
          "value": "span_1",
        },
      ],
      "time": 683273000,
      "value": 683273000,
    },
  ],
}
`;

exports[`utils/spanToTree should return a single tree on valid trace data 2`] = `
Object {
  "missingSpanTree": Array [
    Object {
      "children": Array [
        Object {
          "childReferences": Array [
            Object {
              "RefType": "CHILD_OF",
              "SpanId": "span_2",
              "TraceId": "0000000000000000span_1",
            },
          ],
          "children": Array [],
          "event": Array [
            Object {
              "attributeMap": Object {
                "event": "HTTP request received S3",
                "level": "info",
                "method": "GET",
                "url": "/dispatch?customer=392&nonse=0.015296363321630757",
              },
              "timeUnixNano": 1657275433246142000,
            },
          ],
          "hasError": false,
          "id": "span_3",
          "isProcessed": true,
          "name": "HTTP GET SPAN 3",
          "nonChildReferences": Array [],
          "serviceColour": "",
          "serviceName": "frontend",
          "startTime": 1657275433246,
          "tags": Array [
            Object {
              "key": "host.name.span3",
              "value": "span_3",
            },
          ],
          "time": 683273000,
          "value": 683273000,
        },
      ],
      "id": "span_2",
      "isMissing": true,
      "name": "Missing Span (span_2)",
      "serviceColour": "",
      "serviceName": "",
      "startTime": null,
      "tags": Array [],
      "time": null,
      "value": null,
    },
  ],
  "spanTree": Array [
    Object {
      "childReferences": Array [
        Object {
          "RefType": "CHILD_OF",
          "SpanId": "",
          "TraceId": "0000000000000000span_1",
        },
      ],
      "children": Array [],
      "event": Array [
        Object {
          "attributeMap": Object {
            "event": "HTTP request received S1",
            "level": "info",
            "method": "GET",
            "url": "/dispatch?customer=392&nonse=0.015296363321630757",
          },
          "timeUnixNano": 1657275433246142000,
        },
      ],
      "hasError": false,
      "id": "span_1",
      "name": "HTTP GET SPAN 1",
      "nonChildReferences": Array [],
      "serviceColour": "",
      "serviceName": "frontend",
      "startTime": 1657275433246,
      "tags": Array [
        Object {
          "key": "host.name.span1",
          "value": "span_1",
        },
      ],
      "time": 683273000,
      "value": 683273000,
    },
  ],
}
`;
